src/settingsholder.cpp

```diff
@@ -200,3 +225,101 @@ void SettingsHolder::setAddonSetting(const AddonSettingQuery& query,
     emit addonSettingsChanged();
   }
 }
+
+QString SettingsHolder::journalSettingFileName() const {
+  return QDir(
+#ifdef MVPN_WASM
+             // https://wiki.qt.io/Qt_for_WebAssembly#Files_and_local_file_system_access
+             "/"
+#elif defined(UNIT_TEST)
+             QStandardPaths::writableLocation(QStandardPaths::TempLocation)
+#else
+             QStandardPaths::writableLocation(QStandardPaths::AppDataLocation)
+#endif
+             )
+      .filePath("settings-journal");
+}
+
+bool SettingsHolder::beginTransaction() {
+  if (m_settingsJournal) {
+    logger.warning() << "Nested transactions are not supported";
+    return false;
+  }
+
+  sync();
+
+  if (!QFile::copy(m_settings.fileName(), journalSettingFileName())) {
+    logger.warning() << "Unable to generate a setting journal file"
+                     << journalSettingFileName();
+    return false;
+  }
+
+  m_settingsJournal =
+      new QSettings(journalSettingFileName(), m_settings.format(), this);
+  return true;
+}
+
+bool SettingsHolder::commitTransaction() {
+  if (!m_settingsJournal) {
+    logger.warning() << "We are not in a transaction";
+    return false;
+  }
+
+  return finalizeTransaction();
+}
+
+bool SettingsHolder::rollbackTransaction() {
+  if (!m_settingsJournal) {
+    logger.warning() << "We are not in a transaction";
+    return false;
+  }
+
+  QMap<QString, QPair<const char*, QVariant>> transactionChanges(
+      m_transactionChanges);
+
+  if (!finalizeTransaction()) {
+    return false;
+  }
+
+  QMapIterator<QString, QPair<const char*, QVariant>> i(transactionChanges);
+  while (i.hasNext()) {
+    i.next();
+    m_settings.setValue(i.key(), i.value().second);
+    QMetaObject::invokeMethod(this, i.value().first, Qt::DirectConnection);
+  }
+
+  return true;
+}
+
+bool SettingsHolder::finalizeTransaction() {
+  m_transactionChanges.clear();
+  delete m_settingsJournal;
+  m_settingsJournal = nullptr;
+
+  if (!QFile::remove(journalSettingFileName())) {
+    logger.warning() << "Unable to remove the setting journal file"
```

Comment:
Is a warning really enough here? What if we can't remove and then start a new transaction?

Reply:
Because it sounds like a bug having 2 nested transactions or canceling the previous one.

Reply:
I would add a Q_ASSERT() to prevent this from happening. So far, this is used only by the Tutorial engine and we cannot have 2 tutorials happening at the same time.

Reply:
Hm, but the issue is not really if two tutorials happen at same time. It's if one tutorial happens and cannot delete the journal file, leaves it there corrupted and then another one starts and has to deal with the corrupted file.

Reply:
Good point. I added a telemetry event to monitor how often it happens.